p2p/src/cybershake.rs

```diff
@@ -218,73 +240,255 @@ where
+                    _ => return Err(Error::IoError(e)),
+                },
+            };
+        }
+        self.flush().await.map_err(|e| Error::IoError(e))?;
+        Ok(())
+    }
+}
+
+impl<W: AsyncWrite + Unpin> Outgoing<W> {
+    fn cipher_buf(&mut self) {
         self.kdf.append_u64(b"seq", self.seq);
         let mut key = [0u8; 32];
         self.kdf.challenge_bytes(b"key", &mut key);
 
         let ad = encode_u64le(self.seq);
 
-        let ciphertext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
-            .encrypt(&[&ad], msg)
-            .map_err(|_| Error::ProtocolError)?;
+        self.ciphertext_buf.clear();
 
+        let ciphertext = Aes128PmacSiv::new(GenericArray::clone_from_slice(&key))
+            .encrypt(&[&ad], &self.plaintext_buf)
+            .map_err(|_| unimplemented!())
+            .unwrap();
+
+        // Write length of the message.
+        Write::write(
+            &mut self.ciphertext_buf,
+            &encode_u16le(ciphertext.len() as u16)[..],
+        )
+        .unwrap();
+        // Write a message.
+        Write::write(&mut self.ciphertext_buf, &ciphertext).unwrap();
+
+        self.plaintext_buf.clear();
         self.seq += 1;
+    }
 
-        // Write the length prefix and the ciphertext.
-        self.writer
-            .write(&encode_u64le(ciphertext.len() as u64)[..])
-            .await?;
-        self.writer.write(&ciphertext[..]).await?;
-        self.writer.flush().await?;
-        Ok(())
+    pub fn flush_write(&mut self, cx: &mut Context) -> Poll<Result<(), io::Error>> {
+        loop {
+            let poll = self
+                .writer
+                .as_mut()
+                .poll_write(cx, &self.ciphertext_buf[self.ciphertext_sent..]);
+            let n = ready!(poll);
+            self.ciphertext_sent += n;
+            if self.ciphertext_sent == self.ciphertext_buf.len() {
+                self.ciphertext_sent = 0;
+                self.ciphertext_buf.clear();
+                if self.plaintext_needs_flushing {
+                    self.cipher_buf();
+                    self.plaintext_needs_flushing = false;
+                }
+                return Poll::Ready(Ok(()));
+            }
+        }
     }
 }
 
-impl<W: AsyncRead + Unpin> Incoming<W> {
-    pub async fn receive_message(&mut self) -> Result<Vec<u8>, Error> {
-        let mut lenbuf = [0u8; 8];
-        let seq = self.seq;
-        self.seq += 1;
-        self.reader.read_exact(&mut lenbuf[..]).await?;
-        let len = LittleEndian::read_u64(&lenbuf) as usize;
+impl<W: AsyncWrite + Unpin> AsyncWrite for Outgoing<W> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let me = self.get_mut();
 
-        // length must include IV prefix (16 bytes)
-        if len < 16 {
-            return Err(Error::ProtocolError);
+        if me.plaintext_needs_flushing {
```

Comment:
In that case no more need in two buffers. We can write plaintext in buffer, and then write ct in the same buffer. But i think it will be more slowly, because we need to wait while ct will be sent.

Reply:
Yeah, that'd need one buffer. I think this microlag is okay to have in this initial version, after all, it's a backpressure. Double-buffering does not really reduce average latency, but allows smoothing the peaks if we are doing some sort of continuous streaming, which we certainly do not do here.

Reply:
Reworked